### PR TITLE
Prevent egg hatching in claims without container permission

### LINKED ISSUE #201: [Request]: disable the 'use' (throwing / spawning) of chickens eggs in someone's claim
Feature Request:

Hi, i already asked this on Spigot and you told me that yo would look in to the idea, but asked me to post the 'feature request' here so it won't be forgotten. so here it is:

Please,  add the feature to disable egg bombing on a players claim, and please disable it on default.

I run a shop and everytime player come in to the shop they egg-bomb it, throwing a full inventory of chicken eggs in to my shop spawning chicken.
I find that funny and can take a joke, but not when i have to kill around 2000 chickens (no joke) per 24 hours.

Please add this feature to Grief Prevention

(p.s. i think it's indeed possible to track the 'egg' source and to disallow it. but you'll probably know more about that then me).

### LINKED ISSUE #715: Egg Cannons
One of the many ways we've discovered that is not stopped by GriefPrevention that I don't think is a bug, so filing this as a Feature Request:

Players can build an array of egg-filled dispensers near a claim, overlooking it, then redstone clock mass fire them.   Very quickly, hundreds of chickens.

Would it be possible to intercept egg hatching inside a claim?

### PR DESCRIPTION
**Pull request by YanisBft**
This PR adds a config option to globally toggle egg hatching, similarly to the ender pearl option.

If a player throws a chicken egg in a claim, it will be consumed, but the egg cannot hatch, unless the player is container-trusted in the claim.

Egg hatching is disabled by default.

Closes #201 
Closes #715

### CODE REVIEW

**Comment by Jikoo** [comment]
This looks like it's dependent on the current location of the player when throwing, not where the egg actually breaks. Couldn't players stand in a claim they're trusted in and throw eggs wherever in range?

**Review by RoboMWM — CHANGES_REQUESTED** [review]
Would be better to make this part of the mob spawn eggs option or whatever that is. Either way, this isn't gonna work as per jikoo's comments

**Comment by Jikoo** [comment]
Honestly I don't think it needs a node at all - if you can't breed other things, it stands to reason that you can't spawn chickens.

Re: fixing - should be as simple as converting to ProjectileHitEvent and ensuring the projectile is an egg.

**Comment by YanisBft** [comment]
@RoboMWM So what should I do? Make it part of the spawn egg option and use ProjectileHitEvent?

**Comment by RoboMWM** [comment]
Not 100% sure about the former, but it's the closest mechanic I can think of.

**Comment by RoboMWM** [comment]
is breeding part of containertrust? That's really the only other mechanic I can see this being similar to.

**Comment by Jikoo** [comment]
Breeding is CT, yeah.

**Comment by RoboMWM** [comment]
Hmm, I guess then keep it to containertrust then.

**Comment by YanisBft** [comment]
Changed to ProjectileHitEvent, and reverted the config node addition to only check container permissions

File: src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java [hunk 1]
```diff
@@ -1335,6 +1339,47 @@ public void onPlayerInteractEntity(PlayerInteractEntityEvent event)
...
+                {
+                    message += "  " + instance.dataStore.getMessage(Messages.IgnoreClaimsAdvertisement);
+                }
+
+                GriefPrevention.sendMessage(player, TextMode.Err, message);
+                player.getInventory().addItem(((Egg) event.getEntity()).getItem());
+                return;
+            }
+        }
+    }
```
Comment (RoboMWM) [inline src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java]: Don't see anywhere where you stop the egg from hatching. Am assuming you haven't had an opportunity to test yet.

**Review by RoboMWM — CHANGES_REQUESTED** [review]

File: src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java (hunk 1 above)
Comment (YanisBft) [reply src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java]: Unfortunately this event can't be cancelled. Though I forgot to remove the egg entity, that is now fixed. And yes, I haven't tested it

**Comment by RoboMWM** [comment]
does the hatching happen before or after that event? Because I know you can't cancel that event.

cuz if that works I wonder if I could just remove arrows for the chorus fruit fix I put in.

**Comment by YanisBft** [comment]
So, I tested it and now it works!

ProjectileHitEvent does not work as the egg hatching happens after the hit.

PlayerEggThrowEvent works though, as `event.getEgg().getLocation()` is actually the hit location of the egg. When a player throws an egg that will hit inside a claim (which does not belong to them obv), the hatching is prevented and the player gets back their egg, unless they are not in survival/adventure.

File: src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java [hunk 2]
```diff
@@ -1335,6 +1336,41 @@ public void onPlayerInteractEntity(PlayerInteractEntityEvent event)
...
+            //cancel the event by preventing hatching
+            event.setHatching(false);
+
+            //only give the egg back if player is in survival or adventure
+            if (player.getGameMode() == GameMode.SURVIVAL || player.getGameMode() == GameMode.ADVENTURE)
+            {
+                player.getInventory().addItem(event.getEgg().getItem());
+            }
+
+            return;
```
Comment (RoboMWM) [inline src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java]: I think this return is unnecessary as no other code follows this.

**Comment by RoboMWM** [comment]
Hmm, interesting. Javadocs don't indicate at all exactly when this event occurs. So that's interesting.

**Review by RoboMWM — APPROVED** [review]

**Pull request merged**